### ISSUE TITLE
gen5.9 can't work

### ISSUE THREAD

**Issue by jiamo** [jiamo/combination_generator#1]

**Comment by jiamo** [comment]
✦ The fix works because, in recursive combinatorial generation algorithms (like those based on Gray codes or similar sequences), the neg (negative/reverse) function is expected to perform the inverse operation
  of the gen (generate) function.

  Here's a detailed breakdown:

  1. Understanding gen and neg in this context
   * `gen` (Generate): This function typically generates a sequence of combinations in a "forward" direction.
   * `neg` (Negative/Reverse): This function is meant to generate a sequence of combinations in the "reverse" direction, or effectively "undo" the state changes that gen would make, allowing the algorithm to
     explore other branches of the recursive tree without producing duplicates or missing combinations.

  2. The Base Case k == n - 1
  This specific base case means we are selecting n-1 items out of n, which is equivalent to selecting 1 item not to be included. In the binary res array, this translates to having one 0 and n-1 1s. The loops in
  both gen and neg for this case are essentially moving this single 0 through the array by swapping adjacent elements.

  3. The Original Problem (Before the Fix)
  The original code for the neg function in the base case k == n-1 was:

   1 if k == n - 1:
   2     for i in range(n-1, 0, -1): # Iterates from right to left
   3         swap(i, i+1)
  Crucially, this loop range(n-1, 0, -1) performs swaps in the same order as the corresponding gen function's base case. If gen moves the 0 from right to left, and neg also moves the 0 from right to left, then
  neg is not reversing the operation. Instead, it's just repeating a similar forward step. This leads to the array res being in an incorrect state when neg finishes, causing subsequent recursive calls to gen or
  neg to produce the same combinations again (duplicates) or miss valid combinations.

  4. The Solution (After the Fix)
  The change modified the neg function's base case k == n-1 to:

   1 if k == n - 1:
   2     for i in range(1, n):      # Iterates from left to right
   3         swap(i, i+1)
  By changing the loop to range(1, n), the swap operations are now performed in the opposite order compared to the gen function's base case. If gen "bubbles" the 0 from right to left, this neg now "bubbles" the
  0 from left to right. This ensures that:
   * gen performs a sequence of state transformations: $S_{initial} \xrightarrow{\text{gen steps}} S_{final}$.
   * neg now performs the inverse sequence of state transformations, effectively bringing the array back to a state that allows the algorithm to explore distinct paths in the recursive tree.

  In summary, the modification ensures that neg properly reverses the state changes caused by gen in that specific base case, which is fundamental for correct and non-repetitive generation of combinations in
  such recursive algorithms.
